roachtest: sqlsmith/setup=rand-tables/setting=default failed

roachtest.sqlsmith/setup=rand-tables/setting=default [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2920165&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2920165&tab=artifacts#/sqlsmith/setup=rand-tables/setting=default) on master @ [3cd29547884feb24b150b67ca2a88073798a35fd](https://github.com/cockroachdb/cockroach/commits/3cd29547884feb24b150b67ca2a88073798a35fd):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=default/run_1
	sqlsmith.go:248,sqlsmith.go:262,test_runner.go:777: ping: dial tcp 34.68.31.39:26257: connect: connection refused
		previous sql:
		DELETE FROM
			defaultdb.public.table3 AS tab_338
		WHERE
			tab_338.col3_2
		ORDER BY
			tab_338.col3_4 ASC, tab_338.col3_4 DESC, tab_338.crdb_internal_mvcc_timestamp DESC
		LIMIT
			49:::INT8;

	cluster.go:1716,context.go:140,cluster.go:1705,test_runner.go:858: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2920165-1619291008-06-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 2: 10939
		3: 10356
		4: 10090
		1: dead
		Error: UNCLASSIFIED_PROBLEM: 1: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1162
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:284
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:856
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:960
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:897
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:2057
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:204
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (3) 1: dead
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh sqlsmith/setup=rand-tables/setting=default
```
</p>
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=default.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xeef862]

goroutine 2657 [running]:
panic(0x486a0a0, 0xc004292810)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc003d02b60 sp=0xc003d02a98 pc=0x48b725
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0024f8000, 0x5b6ba20, 0xc0021b9700, 0x4613c80, 0x7e301a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:814 +0x3a9 fp=0xc003d02be0 sp=0xc003d02b60 pc=0x3537ac9
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc0024f8000, 0x5b6ba20, 0xc0021b9700)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:482 +0x65 fp=0xc003d02c18 sp=0xc003d02be0 pc=0x370e4a5
runtime.call32(0x0, 0x53eddd8, 0xc0028a85e8, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:540 +0x3e fp=0xc003d02c48 sp=0xc003d02c18 pc=0x4c30be
runtime.reflectcallSave(0xc003d02d88, 0x53eddd8, 0xc0028a85e8, 0x18)
	/usr/local/go/src/runtime/panic.go:881 +0x58 fp=0xc003d02c78 sp=0xc003d02c48 pc=0x48b118
runtime.runOpenDeferFrame(0xc0005dd680, 0xc0028a85a0, 0x0)
	/usr/local/go/src/runtime/panic.go:855 +0x2cd fp=0xc003d02d08 sp=0xc003d02c78 pc=0x48afcd
panic(0x4613c80, 0x7e301a0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9 fp=0xc003d02dd0 sp=0xc003d02d08 pc=0x48b399
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1(0xc003d03748)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x68b fp=0xc003d02f70 sp=0xc003d02dd0 pc=0x1d187cb
runtime.call32(0x0, 0x53eed80, 0xc0010e5448, 0x800000008)
	/usr/local/go/src/runtime/asm_amd64.s:540 +0x3e fp=0xc003d02fa0 sp=0xc003d02f70 pc=0x4c30be
runtime.reflectcallSave(0xc003d030e0, 0x53eed80, 0xc0010e5448, 0xc000000008)
	/usr/local/go/src/runtime/panic.go:881 +0x58 fp=0xc003d02fd0 sp=0xc003d02fa0 pc=0x48b118
runtime.runOpenDeferFrame(0xc0005dd680, 0xc0010e5400, 0xc003d03128)
	/usr/local/go/src/runtime/panic.go:855 +0x2cd fp=0xc003d03060 sp=0xc003d02fd0 pc=0x48afcd
panic(0x4613c80, 0x7e301a0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9 fp=0xc003d03128 sp=0xc003d03060 pc=0x48b399
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:212
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:742 +0x413 fp=0xc003d03158 sp=0xc003d03128 pc=0x4a2473
github.com/cockroachdb/cockroach/pkg/util/tracing.SpanFromContext(0x0, 0x0, 0xd8)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/context.go:25 +0x22 fp=0xc003d03190 sp=0xc003d03158 pc=0xeef862
github.com/cockroachdb/cockroach/pkg/sql/execinfra.GetTraceData(0x0, 0x0, 0xd8, 0x7f1d1d3be7d0, 0xc0040f2d20)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:243 +0x35 fp=0xc003d031c0 sp=0xc003d03190 pc=0x245c3f5
github.com/cockroachdb/cockroach/pkg/sql/execinfra.GetCumulativeContentionTime(0x0, 0x0, 0xc0011cf450)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/stats.go:35 +0x45 fp=0xc003d03238 sp=0xc003d031c0 pc=0x2465405
github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*ColBatchScan).GetCumulativeContentionTime(0xc0011cf400, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/colbatch_scan.go:166 +0x37 fp=0xc003d03260 sp=0xc003d03238 pc=0x337ea57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedStatsCollectorImpl).GetStats(0xc0040496c0, 0xc0040d3d18)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:188 +0x26d fp=0xc003d032d0 sp=0xc003d03260 pc=0x339dccd
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*drainHelper).Next(0xc002019aa0, 0x12e5496, 0x4ce9dd9, 0x4ce9dd9, 0x5)
...
```

Second hit is:
```
panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3

goroutine 2805 [running]:
panic(0x486a560, 0xc00252b350)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc003d234e0 sp=0xc003d23418 pc=0x48b725
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0011c1900, 0x5b6c160, 0xc001beb700, 0x4a40960, 0xc002b171a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:814 +0x3a9 fp=0xc003d23560 sp=0xc003d234e0 pc=0x3538309
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc0011c1900, 0x5b6c160, 0xc001beb700)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:482 +0x65 fp=0xc003d23598 sp=0xc003d23560 pc=0x370ed85
runtime.call32(0x0, 0x53ee3e0, 0xc000146288, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:540 +0x3e fp=0xc003d235c8 sp=0xc003d23598 pc=0x4c30be
runtime.reflectcallSave(0xc003d23708, 0x53ee3e0, 0xc000146288, 0x18)
	/usr/local/go/src/runtime/panic.go:881 +0x58 fp=0xc003d235f8 sp=0xc003d235c8 pc=0x48b118
runtime.runOpenDeferFrame(0xc0011c4300, 0xc000146240, 0x0)
	/usr/local/go/src/runtime/panic.go:855 +0x2cd fp=0xc003d23688 sp=0xc003d235f8 pc=0x48afcd
panic(0x4a40960, 0xc002b171a0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9 fp=0xc003d23750 sp=0xc003d23688 pc=0x48b399
runtime.goPanicIndex(0x3, 0x3)
	/usr/local/go/src/runtime/panic.go:88 +0xa5 fp=0xc003d23798 sp=0xc003d23750 pc=0x489225
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).bufferRow(0xc0010f5600, 0x5b8a620, 0xc0034bee78, 0xc002db9fc0, 0x4, 0x4, 0x0, 0x0, 0x0, 0x200000000, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:1238 +0x3f0 fp=0xc003d23820 sp=0xc003d23798 pc=0x3bfdef0
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*commandResult).AddRow(0xc0010f5c80, 0x5b8a620, 0xc0034bee78, 0xc002db9fc0, 0x4, 0x4, 0x4, 0x3)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/command_result.go:206 +0x19b fp=0xc003d238c0 sp=0xc003d23820 pc=0x3bf675b
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLReceiver).Push(0xc003913880, 0xc002fcc240, 0x4, 0x4, 0x0, 0xc0034bf358)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:740 +0x8f6 fp=0xc003d23a20 sp=0xc003d238c0 pc=0x35bd2b6
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run(0x5b8a620, 0xc0034bf358, 0x5b90fe0, 0xc003441800, 0x5b34420, 0xc003913880)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:178 +0x5a fp=0xc003d23a60 sp=0xc003d23a20 pc=0x245c8ba
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run(0xc003441800, 0x5b6c160, 0xc002db92c0)
...
```
which is a dup of #64181.

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -550,6 +550,23 @@ func (r opResult) makeDiskBackedSorterConstructor(
 	}
 }
 
+// TODO(yuzefovich): introduce some way to unit test that the meta info tracking
+// works correctly. See #64256 for more details.
+
+// takeOverMetaInfo iterates over all meta components from the input trees and
+// passes the responsibility of handling those to target. The input objects are
+// updated in-place accordingly.
+func takeOverMetaInfo(target *colexecargs.OpWithMetaInfo, inputs []colexecargs.OpWithMetaInfo) {
+	for i := range inputs {
+		target.StatsCollectors = append(target.StatsCollectors, inputs[i].StatsCollectors...)
+		target.MetadataSources = append(target.MetadataSources, inputs[i].MetadataSources...)
+		target.ToClose = append(target.ToClose, inputs[i].ToClose...)
+		inputs[i].MetadataSources = nil
+		inputs[i].StatsCollectors = nil
+		inputs[i].ToClose = nil
+	}
+}
+
 // createAndWrapRowSource takes a processor spec, creating the row source and
 // wrapping it using wrapRowSources. Note that the post process spec is included
 // in the processor creation, so make sure to clear it if it will be inspected
@@ -628,6 +645,7 @@ func (r opResult) createAndWrapRowSource(
 	if args.TestingKnobs.PlanInvariantsCheckers {
 		r.Root = colexec.NewInvariantsChecker(r.Root)
 	}
+	takeOverMetaInfo(&r.OpWithMetaInfo, inputs)
 	r.MetadataSources = append(r.MetadataSources, r.Root.(colexecop.MetadataSource))
 	r.ToClose = append(r.ToClose, c)
 	r.Releasables = append(r.Releasables, releasables...)
@@ -1404,14 +1422,7 @@ func NewColOperator(
 			r.Root = colexec.NewInvariantsChecker(r.Root)
 		}
 	}
-	// Handle the metadata components from the input trees. Note that it is
-	// possible that we have created materializers which took over the
-	// responsibility over those objects.
-	for i := range inputs {
-		r.StatsCollectors = append(r.StatsCollectors, inputs[i].StatsCollectors...)
-		r.MetadataSources = append(r.MetadataSources, inputs[i].MetadataSources...)
-		r.ToClose = append(r.ToClose, inputs[i].ToClose...)
-	}
+	takeOverMetaInfo(&result.OpWithMetaInfo, inputs)
 	if util.CrdbTestBuild {
 		r.AssertInvariants()
 	}
@@ -1452,9 +1463,11 @@ func (r opResult) planAndMaybeWrapFilter(
 			ProcessorID: processorID,
 			ResultTypes: args.Spec.ResultTypes,
 		}
+		inputToMaterializer := colexecargs.OpWithMetaInfo{Root: r.Root}
+		takeOverMetaInfo(&inputToMaterializer, args.Inputs)
 		return r.createAndWrapRowSource(
-			ctx, flowCtx, args, []colexecargs.OpWithMetaInfo{r.OpWithMetaInfo}, [][]*types.T{r.ColumnTypes},
-			filtererSpec, factory, err,
+			ctx, flowCtx, args, []colexecargs.OpWithMetaInfo{inputToMaterializer},
+			[][]*types.T{r.ColumnTypes}, filtererSpec, factory, err,
 		)
 	}
 	r.Root = op
@@ -1482,9 +1495,11 @@ func (r opResult) wrapPostProcessSpec(
 		Post:        *post,
 		ResultTypes: resultTypes,
 	}
+	inputToMaterializer := colexecargs.OpWithMetaInfo{Root: r.Root}
+	takeOverMetaInfo(&inputToMaterializer, args.Inputs)
 	return r.createAndWrapRowSource(
-		ctx, flowCtx, args, []colexecargs.OpWithMetaInfo{r.OpWithMetaInfo}, [][]*types.T{r.ColumnTypes},
-		noopSpec, factory, causeToWrap,
+		ctx, flowCtx, args, []colexecargs.OpWithMetaInfo{inputToMaterializer},
+		[][]*types.T{r.ColumnTypes}, noopSpec, factory, causeToWrap,
 	)
 }
 

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -1433,13 +1433,13 @@ func (r opResult) planAndMaybeWrapFilter(
 		ctx, flowCtx, evalCtx, r.Root, r.ColumnTypes, filter, args.StreamingMemAccount, factory, args.ExprHelper,
 	)
 	if err != nil {
-		// ON expression planning failed. Fall back to planning the filter
+		// Filter expression planning failed. Fall back to planning the filter
 		// using row execution.
 		if log.V(2) {
 			log.Infof(
 				ctx,
-				"vectorized join ON expr planning failed with error %v ON expr is %s, attempting to wrap as a row source",
-				err, filter.String(),
+				"filter expr %s planning failed with error, attempting to wrap as a row source: %v",
+				filter.String(), err,
 			)
 		}
 

--- a/pkg/sql/colexec/columnarizer.go
+++ b/pkg/sql/colexec/columnarizer.go
@@ -177,12 +177,13 @@ func (c *Columnarizer) GetStats() *execinfrapb.ComponentStats {
 			"unexpectedly the columnarizer was removed from the flow when stats are being collected",
 		))
 	}
-	if !c.removedFromFlow && c.ExecStatsForTrace != nil {
-		s := c.ExecStatsForTrace()
-		s.Component = c.FlowCtx.ProcessorComponentID(c.ProcessorID)
-		return s
+	componentID := c.FlowCtx.ProcessorComponentID(c.ProcessorID)
+	if c.removedFromFlow || c.ExecStatsForTrace == nil {
+		return &execinfrapb.ComponentStats{Component: componentID}
 	}
-	return nil
+	s := c.ExecStatsForTrace()
+	s.Component = componentID
+	return s
 }
 
 // Next is part of the Operator interface.

--- a/pkg/sql/colexecop/operator.go
+++ b/pkg/sql/colexecop/operator.go
@@ -398,6 +398,7 @@ func (s MetadataSources) DrainMeta(ctx context.Context) []execinfrapb.ProducerMe
 // "network" option (strictly for colrpc.Inboxes).
 type VectorizedStatsCollector interface {
 	Operator
-	// GetStats returns the execution statistics of a single Operator.
+	// GetStats returns the execution statistics of a single Operator. It will
+	// always return non-nil (but possibly empty) object.
 	GetStats() *execinfrapb.ComponentStats
 }

--- a/pkg/sql/colflow/stats.go
+++ b/pkg/sql/colflow/stats.go
@@ -44,16 +44,17 @@ type batchInfoCollector struct {
 	componentID execinfrapb.ComponentID
 
 	mu struct {
-		// We need a mutex because finish() and Next() might be called from
-		// different goroutines.
+		// We need a mutex because finishAndGetStats() and Next() might be
+		// called from different goroutines.
 		syncutil.Mutex
+		initialized           bool
 		numBatches, numTuples uint64
 	}
 
 	// stopwatch keeps track of the amount of time the wrapped operator spent
 	// doing work. Note that this will include all of the time that the operator's
 	// inputs spent doing work - this will be corrected when stats are reported
-	// in finish().
+	// in finishAndGetStats().
 	stopwatch *timeutil.StopWatch
 
 	// childStatsCollectors contains the stats collectors for all of the inputs
@@ -80,7 +81,17 @@ func makeBatchInfoCollector(
 	}
 }
 
-// Next is part of the Operator interface.
+// Init is part of the colexecop.Operator interface.
+func (bic *batchInfoCollector) Init(ctx context.Context) {
+	bic.Operator.Init(ctx)
+	bic.mu.Lock()
+	// If we got here, then Init above succeeded, so the wrapped operator has
+	// been properly initialized.
+	bic.mu.initialized = true
+	bic.mu.Unlock()
+}
+
+// Next is part of the colexecop.Operator interface.
 func (bic *batchInfoCollector) Next() coldata.Batch {
 	var batch coldata.Batch
 	bic.stopwatch.Start()
@@ -95,8 +106,13 @@ func (bic *batchInfoCollector) Next() coldata.Batch {
 	return batch
 }
 
-// finish calculates the final statistics.
-func (bic *batchInfoCollector) finish() (numBatches, numTuples uint64, time time.Duration) {
+// finishAndGetStats calculates the final execution statistics for the wrapped
+// operator. ok indicates whether the stats collection was successful.
+func (bic *batchInfoCollector) finishAndGetStats() (
+	numBatches, numTuples uint64,
+	time time.Duration,
+	ok bool,
+) {
 	tm := bic.stopwatch.Elapsed()
 	// Subtract the time spent in each of the child stats collectors, to produce
 	// the amount of time that the wrapped operator spent doing work itself, not
@@ -106,7 +122,7 @@ func (bic *batchInfoCollector) finish() (numBatches, numTuples uint64, time time
 	}
 	bic.mu.Lock()
 	defer bic.mu.Unlock()
-	return bic.mu.numBatches, bic.mu.numTuples, tm
+	return bic.mu.numBatches, bic.mu.numTuples, tm, bic.mu.initialized
 }
 
 // getElapsedTime implements the childStatsCollector interface.
@@ -153,16 +169,26 @@ type vectorizedStatsCollectorImpl struct {
 
 // GetStats is part of the colexecop.VectorizedStatsCollector interface.
 func (vsc *vectorizedStatsCollectorImpl) GetStats() *execinfrapb.ComponentStats {
-	numBatches, numTuples, time := vsc.batchInfoCollector.finish()
+	numBatches, numTuples, time, ok := vsc.batchInfoCollector.finishAndGetStats()
+	if !ok {
+		// The stats collection wasn't successful for some reason, so we will
+		// return an empty object (since nil is not allowed by the contract of
+		// GetStats).
+		//
+		// Such scenario can occur, for example, if the operator wrapped by the
+		// batchInfoCollector wasn't properly initialized, yet the stats
+		// retrieval is attempted. In many places we assume that if an operator
+		// is interacted with in any way, it must have been successfully
+		// initialized. Having such a check in the vectorizedStatsCollectorImpl
+		// makes sure that we never violate those assumptions.
+		return &execinfrapb.ComponentStats{}
+	}
 
 	var s *execinfrapb.ComponentStats
 	if vsc.columnarizer != nil {
 		s = vsc.columnarizer.GetStats()
-	}
-	if s == nil {
-		// Either there was no root columnarizer or it has been
-		// removed from the flow (in which case the columnarizer will return
-		// nil). Create a new stats object.
+	} else {
+		// There was no root columnarizer, so create a new stats object.
 		s = &execinfrapb.ComponentStats{Component: vsc.componentID}
 	}
 
@@ -223,7 +249,20 @@ type networkVectorizedStatsCollectorImpl struct {
 
 // GetStats is part of the colexecop.VectorizedStatsCollector interface.
 func (nvsc *networkVectorizedStatsCollectorImpl) GetStats() *execinfrapb.ComponentStats {
-	numBatches, numTuples, time := nvsc.batchInfoCollector.finish()
+	numBatches, numTuples, time, ok := nvsc.batchInfoCollector.finishAndGetStats()
+	if !ok {
+		// The stats collection wasn't successful for some reason, so we will
+		// return an empty object (since nil is not allowed by the contract of
+		// GetStats).
+		//
+		// Such scenario can occur, for example, if the operator wrapped by the
+		// batchInfoCollector wasn't properly initialized, yet the stats
+		// retrieval is attempted. In many places we assume that if an operator
+		// is interacted with in any way, it must have been successfully
+		// initialized. Having such a check in the vectorizedStatsCollectorImpl
+		// makes sure that we never violate those assumptions.
+		return &execinfrapb.ComponentStats{}
+	}
 
 	s := &execinfrapb.ComponentStats{Component: nvsc.componentID}
 

